### ISSUE TITLE
Enhancement - Add timestamp for start of request

### ISSUE THREAD

**Issue by Tuxino** [marcobreveglieri/dmvc-prometheus-metrics#3]
Its useful to be able to generate elapsed request time for successful and unsuccessful requests.

For a successful request, we can write event handlers for OnBeforeAction and OnAfterAction to determine elapsed time and add this to a Prometheus Histogram. However, if an error occurs, its possible that neither of these event handlers get called eg. for a 404.

I propose that we add a timestamp to the existing DMVC middleware

```
procedure TMetricsMiddleware.OnBeforeRouting(Context: TWebContext; var Handled: Boolean);
begin
  Context.Data.AddOrSetValue('RequestStartTime', FloatToStr(Now));
  // Check whether the current path request matches the metrics one.
  if not SameText(Context.Request.PathInfo, FPathInfo) then
    Exit;
...
```

This enables either the OnAfterAction or a global exception handler to figure out the elapsed time and add it to a histogram.

**Comment by marcobreveglieri** [comment]
Hi @Tuxino,
thanks for your suggestion, I think it's really a good idea!

Would you be able to provide a simple pull request with a code sample to propose and evaluate?

It would be very much appreciated to have a real scenario to evaluate and possibly integrate into the main branch.

Thanks for your feedback! 💘
Marco.

**Comment by Tuxino** [comment]
Hi Marco,

Sorry for the delay - I have created a pull request for this change.

https://github.com/marcobreveglieri/dmvc-prometheus-metrics/pull/4

I am a bit new to Github so please let me know if I have done this correctly.

Thanks

David

**Comment by marcobreveglieri** [comment]
> Hi Marco,
> Sorry for the delay - I have created a pull request for this change. [...]
> I am a bit new to Github so please let me know if I have done this correctly.

Today I've found some time to examine your pull request for this issue. 🤗

I wrote an implementation starting from your code and making it available directly inside the middleware class, with some options to configure it and additional checks.

The started demo and readme information have been updated accordingly.

Thanks a lot for your suggestion, your code and your contribute, really appreciated! 💘

Marco.
